R/read-write.R

```diff
@@ -0,0 +1,428 @@
+
+  table_tibble <- map(
+    table_sheets_sorted,
+    readxl::read_xlsx, path = xlsx_file_path
+  ) %>%
+    set_names(table_names) %>%
+    enframe(name = "table", value = "data")
+
+  # since reading from xlsx via `readxl::read_xlsx()` does e.g. not support reading as integer
+  # we reclass the columns in retrospect:
+  table_tibble_reclassed <- left_join(table_tibble, col_class_info, by = "table") %>%
+    select(-table) %>%
+    pmap(function(data, column, class) {
+      reduce2(column, class, function(data, column, class) {
+      mutate(data, !!column := get(paste0("as.", class))(!!sym(column)))
+    }, .init = data)
+    }) %>%
+    set_names(table_names) %>%
+    enframe(name = "table", value = "data")
+
+  make_dm(def_base, table_names, table_tibble_reclassed, pk_info, fk_info)
+}
+
+make_dm <- function(def_base, table_names, table_tibble, pk_info, fk_info) {
+
+  pk_tibble <- pk_info %>%
+    group_by(table) %>%
+    mutate(pks = list(tibble(column = list(pk_col)))) %>%
+    ungroup() %>%
+    select(-pk_col)
+
+  fk_tibble <- fk_info %>%
+    group_by(child_table, key_nr) %>%
+    summarize(fks = list(
+      tibble(
+        table = unique(parent_table),
+        column = list(fk_col)
+      )
+    ), .groups = "drop_last"
+    ) %>%
+    summarize(
+      fks = list(bind_rows(fks)),
+      .groups = "drop"
+    )
+
+  zoom <- new_zoom()
+  col_tracker_zoom <- new_col_tracker_zoom()
+  filter_tibble <-
+    tibble(
+      table = table_names,
+      filters = vctrs::list_of(new_filter())
+    )
+
+  def_base %>%
+    left_join(table_tibble, by = "table") %>%
+    relocate(data, .after = "table") %>%
+    left_join(pk_tibble, by = "table") %>%
+    mutate(pks = map(pks, ~ if (is_null(.x)) {new_pk()} else {.x})) %>%
+    mutate(pks = vctrs::as_list_of(pks)) %>%
+    left_join(fk_tibble, by = c("table" = "child_table")) %>%
+    mutate(fks = map(fks, ~ if (is_null(.x)) {new_fk()} else {.x})) %>%
+    mutate(fks = vctrs::as_list_of(fks)) %>%
+    left_join(filter_tibble, by = "table") %>%
+    left_join(zoom, by = "table") %>%
+    left_join(col_tracker_zoom, by = "table") %>%
+    new_dm3()
+}
+
+prepare_tables <- function(dm) {
+  check_param_class(dm, "dm")
+  check_not_zoomed(dm, -3)
+  check_no_filter(dm, -3)
+  if (is_empty(dm)) {abort_empty_dm()}
+
+  if (is_db(dm_get_src(dm))) {
+    abort_only_for_local_src(dm_get_src(dm))
+  }
+
+  info_tibble <- dm_get_def(dm) %>% select(table, segment, display)
+
+  col_class_tibble <- dm_col_class(dm)
+  if ("list" %in% col_class_tibble$class) {
+    tbl_col_list <- filter(col_class_tibble, class == "list")
+    abort_no_list(tbl_col_list$table, tbl_col_list$column)
+  }
+  if ("datetime" %in% col_class_tibble$class |
+      "POSIXct" %in% col_class_tibble$class |
+      "POSIXlt" %in% col_class_tibble$class) {
+    tbl_dt_list <- filter(
+      col_class_tibble,
+      class == "datetime" |
+        class == "POSIXct" |
+        class == "POSIXlt"
+    )
+    message(
+      glue::glue(
```

Comment:
The conversion to UTC is lossless, we can convert ourselves (and print the timestamp as `'....+00:00'` or `'...Z'`.